### ISSUE TITLE
[JENKINS-13531] Plugin replacing "+" with " " in configuration strings when plugin is instantiated.

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/jenkins#12644]
<p>I've been experimenting with the Amazon S3 Publisher plugin in Jenkins 1.460 in preparation for starting to use S3 for artifact storage & program distribution at work. I kept getting errors with the S3 plugin, however, stating "Can't connect to S3 service: The request signature we calculated does not match the signature you provided. Check your key and signing method."  </p>

<p>The Access & Secret Keys were correct and being stored correctly in the hudson.plugins.s3.S3BucketPublisher.xml configuration file.  I added some logging to the plugin to discover that in S3BucketPublisher.DescriptorImpl.doLoginCheck(), the "secretKey" element of the StaplerRequest parameter was being returned incorrectly.  There's a "+" character in the secret key.  The plus was being turn into a space (" "), thus the plugin is unable to connect to S3.</p>

<p>The issue first appears with Jenkins & the S3 Publisher plugin in Jenkins 1.455 and continues through 1.460.  Versions 1.454 and prior behave as expected.  The "+" in the secret key is retained and connection to S3 is possible. Nothing has changed in the S3 plugin in that time period, so the issue must be somewhere inside Jenkins itself.  Unfortunately, I'm rather unfamiliar with the Jenkins architecture and plugin architecture an am unable to trace the issue further down the chain than that.</p>

<p>To recreate the issue:</p>

<p>1) get the S3 plugin (<a href="https://github.com/jenkinsci/s3-plugin" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/s3-plugin</a>)<br/>
2) set the jenkins version on line 6 of pom.xml to 1.455 or greater.<br/>
3) in "Configure System", add an S3 profile.  Valid or not does not matter.  Make sure there's a "+" in the secret key or the access key field.<br/>
4) Set a breakpoint, or print out the value of req.getParameter("secretKey") in S3BucketPublisher.DescriptorImpl.doLoginCheck().  See that the "+" has been turned into a " ".  </p>

<p>The strange thing is that if you look in the actual form fields secretKey or accessKey, the + will be in there correctly. Somehow it's not getting to the actual plugin code as a +, though.</p>

<p>Workarounds:</p>

<p>None known at this time.  </p>

<p>I attempted to replace the + with its URLEncoded form "%2B" in the configuration file, but "%2B" comes through instead of being decoded into a +.</p>

<p>The only hack I have to get it working for us at the office for the time being is to replace all instances of " " in the secretKey with "+".  Not a good solution.</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/glimberg.jpg" title="glimberg's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=glimberg">glimberg</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13531" target="_blank">Plugin replacing "+" with " " in configuration strings when plugin is instantiated.</a></i></summary>
<i><ul>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>component(s)</b>: core, plugin-proposals
<li><b>label(s)</b>: configuration, plugins, url-encoding
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2012-09-22T05:48:04+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 4
<li><b>imported</b>: 2025-11-24
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>I've been experimenting with the Amazon S3 Publisher plugin in Jenkins 1.460 in preparation for starting to use S3 for artifact storage &amp; program distribution at work. I kept getting errors with the S3 plugin, however, stating "Can't connect to S3 service: The request signature we calculated does not match the signature you provided. Check your key and signing method."  </p>

<p>The Access &amp; Secret Keys were correct and being stored correctly in the hudson.plugins.s3.S3BucketPublisher.xml configuration file.  I added some logging to the plugin to discover that in S3BucketPublisher.DescriptorImpl.doLoginCheck(), the "secretKey" element of the StaplerRequest parameter was being returned incorrectly.  There's a "+" character in the secret key.  The plus was being turn into a space (" "), thus the plugin is unable to connect to S3.</p>

<p>The issue first appears with Jenkins &amp; the S3 Publisher plugin in Jenkins 1.455 and continues through 1.460.  Versions 1.454 and prior behave as expected.  The "+" in the secret key is retained and connection to S3 is possible. Nothing has changed in the S3 plugin in that time period, so the issue must be somewhere inside Jenkins itself.  Unfortunately, I'm rather unfamiliar with the Jenkins architecture and plugin architecture an am unable to trace the issue further down the chain than that.</p>

<p>To recreate the issue:</p>

<p>1) get the S3 plugin (<a href="https://github.com/jenkinsci/s3-plugin" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/s3-plugin</a>)
2) set the jenkins version on line 6 of pom.xml to 1.455 or greater.
3) in "Configure System", add an S3 profile.  Valid or not does not matter.  Make sure there's a "+" in the secret key or the access key field.
4) Set a breakpoint, or print out the value of req.getParameter("secretKey") in S3BucketPublisher.DescriptorImpl.doLoginCheck().  See that the "+" has been turned into a " ".  </p>

<p>The strange thing is that if you look in the actual form fields secretKey or accessKey, the + will be in there correctly. Somehow it's not getting to the actual plugin code as a +, though.</p>

<p>Workarounds:</p>

<p>None known at this time.  </p>

<p>I attempted to replace the + with its URLEncoded form "%2B" in the configuration file, but "%2B" comes through instead of being decoded into a +.</p>

<p>The only hack I have to get it working for us at the office for the time being is to replace all instances of " " in the secretKey with "+".  Not a good solution.</p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>RedHat Linux and Mac OS X Lion</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>cjo9900:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13531?focusedId=161994&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-161994">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>

<p>Tests on Jenkins 1.460 using firebug show that the Post message is sent incorrectly encoded.</p>

<p>Parametersapplication/x-www-form-urlencoded
accessKey	l ll
name	kik
secretKey	
Source
name=kik&amp;secretKey=&amp;accessKey=l++ll</p>

<p>according to the encoding spec
the query string should be encoded as follows for the plus and space
'+' -&gt; %2B
' ' -&gt; '+'</p>

<p>so should be sent as 
accessKey	l%2B%2Bll</p>

<p>The issue is caused by the change in ./war/src/main/webapp/scripts/prototype.js which changed from version 1.5.1.1 to 1.7 between Jenkins 1.454 and 1.455. The change in this file is that the Ajax.Request class and its base class handles the parameters differently, in 1.5.1.1 a string type parameter was converted to a params hash object including a URIdecode on it. in 1.7 it stores the String parameters directly uses them as is and therefore if it is not correctly URI encoded this gets passed to the backend Winstone/Stapler which calls URIdecode on it accoring to the URI specs for query and converts the '+' to a space.</p>

<p>2 solutions 
Hack prototype.js to perform this encoding</p>

<p>Change the s3 plugin global.jelly to use encodeURIComponent() rather than escape()</p>
</pre>
</details>

</ul>
</details>


<p>Tests on Jenkins 1.460 using firebug show that the Post message is sent incorrectly encoded.</p>

<p>Parametersapplication/x-www-form-urlencoded<br/>
accessKey	l ll<br/>
name	kik<br/>
secretKey	<br/>
Source<br/>
name=kik&secretKey=&accessKey=l++ll</p>

<p>according to the encoding spec<br/>
the query string should be encoded as follows for the plus and space<br/>
'+' -> %2B<br/>
' ' -> '+'</p>

<p>so should be sent as <br/>
accessKey	l%2B%2Bll</p>

<p>The issue is caused by the change in ./war/src/main/webapp/scripts/prototype.js which changed from version 1.5.1.1 to 1.7 between Jenkins 1.454 and 1.455. The change in this file is that the Ajax.Request class and its base class handles the parameters differently, in 1.5.1.1 a string type parameter was converted to a params hash object including a URIdecode on it. in 1.7 it stores the String parameters directly uses them as is and therefore if it is not correctly URI encoded this gets passed to the backend Winstone/Stapler which calls URIdecode on it accoring to the URI specs for query and converts the '+' to a space.</p>

<p>2 solutions <br/>
Hack prototype.js to perform this encoding</p>

<p>Change the s3 plugin global.jelly to use encodeURIComponent() rather than escape()</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/glimberg.jpg" title="glimberg's avatar" /> glimberg:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13531?focusedId=162008&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-162008">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I just verified that your patch works and added a comment to your pull request.  Thanks cjo.</p>
</pre>
</details>

</ul>
</details>

<p>I just verified that your patch works and added a comment to your pull request.  Thanks cjo.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>jlbthetechy:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13531?focusedId=162051&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-162051">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I just experienced the same issue with this plugin. I made the change to the jelly file mentioned and that appeared to work. </p>
</pre>
</details>

</ul>
</details>

<p>I just experienced the same issue with this plugin. I made the change to the jelly file mentioned and that appeared to work. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/glimberg.jpg" title="glimberg's avatar" /> glimberg:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13531?focusedId=162178&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-162178">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>So there's a pull request waiting in the s3-plugin repo.  Any chance on getting this merged into the main line any time soon?</p>

<p><a href="https://github.com/jenkinsci/s3-plugin/pull/4" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/s3-plugin/pull/4</a></p>
</pre>
</details>

</ul>
</details>

<p>So there's a pull request waiting in the s3-plugin repo.  Any chance on getting this merged into the main line any time soon?</p>

<p><a href="https://github.com/jenkinsci/s3-plugin/pull/4" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/s3-plugin/pull/4</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/dbrinker.png" title="dbrinker's avatar" /> dbrinker:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13531?focusedId=167870&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-167870">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I'll second the above.  Any chance this can get merged in soon?</p>
</pre>
</details>

</ul>
</details>

<p>I'll second the above.  Any chance this can get merged in soon?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13531?focusedId=167874&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-167874">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: cjo9900
Path:
 src/main/resources/hudson/plugins/s3/S3BucketPublisher/global.jelly
<a href="http://jenkins-ci.org/commit/s3-plugin/f2caf34a52219a83768616c634279cbe59b59c36" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/s3-plugin/f2caf34a52219a83768616c634279cbe59b59c36</a>
Log:
  Change escape to encodeURIcomponent so that plus chars are converted
correctly for s3 secret and access key</p>

<p><span class="error">&#91;FIXED JENKINS-13531&#93;</span></p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: cjo9900<br/>
Path:<br/>
 src/main/resources/hudson/plugins/s3/S3BucketPublisher/global.jelly<br/>
<a href="http://jenkins-ci.org/commit/s3-plugin/f2caf34a52219a83768616c634279cbe59b59c36" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/s3-plugin/f2caf34a52219a83768616c634279cbe59b59c36</a><br/>
Log:<br/>
  Change escape to encodeURIcomponent so that plus chars are converted<br/>
correctly for s3 secret and access key</p>

<p><span class="error">&#91;FIXED JENKINS-13531&#93;</span></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13531?focusedId=167875&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-167875">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Nicolas De loof
Path:
 src/main/resources/hudson/plugins/s3/S3BucketPublisher/global.jelly
<a href="http://jenkins-ci.org/commit/s3-plugin/ce00365e794f66de5e3ada9633a9ddbcc0454112" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/s3-plugin/ce00365e794f66de5e3ada9633a9ddbcc0454112</a>
Log:
  Merge pull request #4 from cjo9900/<a href="https://issues.jenkins.io/browse/JENKINS-13531" title="Plugin replacing &quot;+&quot; with &quot; &quot; in configuration strings when plugin is instantiated." class="issue-link" data-issue-key="JENKINS-13531"><del>JENKINS-13531</del></a></p>

<p><a href="https://issues.jenkins.io/browse/JENKINS-13531" title="Plugin replacing &quot;+&quot; with &quot; &quot; in configuration strings when plugin is instantiated." class="issue-link" data-issue-key="JENKINS-13531"><del>JENKINS-13531</del></a> Change escape to encodeURIcomponent so that plus chars are converted</p>


<p>Compare: <a href="https://github.com/jenkinsci/s3-plugin/compare/5ab7bde2428f...ce00365e794f" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/s3-plugin/compare/5ab7bde2428f...ce00365e794f</a></p>

</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Nicolas De loof<br/>
Path:<br/>
 src/main/resources/hudson/plugins/s3/S3BucketPublisher/global.jelly<br/>
<a href="http://jenkins-ci.org/commit/s3-plugin/ce00365e794f66de5e3ada9633a9ddbcc0454112" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/s3-plugin/ce00365e794f66de5e3ada9633a9ddbcc0454112</a><br/>
Log:<br/>
  Merge pull request #4 from cjo9900/<a href="https://issues.jenkins.io/browse/JENKINS-13531" title="Plugin replacing "+" with " " in configuration strings when plugin is instantiated." class="issue-link" data-issue-key="JENKINS-13531"><del>JENKINS-13531</del></a></p>

<p><a href="https://issues.jenkins.io/browse/JENKINS-13531" title="Plugin replacing "+" with " " in configuration strings when plugin is instantiated." class="issue-link" data-issue-key="JENKINS-13531"><del>JENKINS-13531</del></a> Change escape to encodeURIcomponent so that plus chars are converted</p>


<p>Compare: <a href="https://github.com/jenkinsci/s3-plugin/compare/5ab7bde2428f...ce00365e794f" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/s3-plugin/compare/5ab7bde2428f...ce00365e794f</a></p>
